### DISCUSSION TITLE
Should ABCI applications deal with concurrency?

### DISCUSSION THREAD

**Discussion by melekes** [tendermint/tendermint#6171] [Q&A]
Right now, all ABCI clients (grpc, socket, local) guarantee that all calls are serialized. The socket server also uses a `sync.Mutex` to serialise calls to the app. Even though we have 4 connections (query, mempool, consensus, statesync), requests coming from Tendermint to the app are serialized https://github.com/tendermint/tendermint/blob/c1ca749afabe7cbe4fd330fbb8dd34ff5b04ac9f/abci/server/socket_server.go#L187.

The question is do we want to loosen the guarantees here and allow some level of concurrency? Or should we offload the concurrency problem to the app?

Related issues:

- https://github.com/tendermint/tendermint/issues/6048

**Comment by jinsankim** [comment]
I think we should offload the concurrency issue to the app. In terms of ABCI, `tendermint` is a client and `app` is a server. In general approach, a server should protect itself from concurrency and also I think it's efficient way. I believe we could achieve fairly big performance gain from it.

**Reply by alexanderbez** [reply]
I think this is a fair line of reasoning.

@melekes it might be helpful to describe why all calls are serialized, specifically the ones that have no bearing on state mutations.

**Reply by melekes** [reply]
> it might be helpful to describe why all calls are serialized

not sure 🤔 I will try to dig in commit & issue history

**Reply by cmwaters** [reply]
At a glance, I would also agree that this makes sense.

**Reply by creachadair** [reply]
I agree that the application should be responsible to control how much parallelism it wants to allow.

I broadly think that one current design feature of ABCI, wherein we use separate connections from the consensus engine to the application, is a mistake, and it muddies the waters between concurrency (a semantic property) and parallelism (an implementation choice). The rationale seems to have been to _allow_ the application to work in parallel, but separate connections are not required for that, and as this issue shows we're not really even making use of that ability. So my general view is that we should either take advantage of the design intent or simplify the interface. (I think we could in fact do both, but there's some inertia here that may not be easily overcome)

Whatever we decide, we need to (more) clearly document presumptions about concurrency in the protocol. The various method groups are clearly _intended_ to function concurrently (even though we don't take advantage of that yet, hence this issue), but it's not clear whether that concurrency is a _necessary_ feature of the API—meaning, for example, that without it consensus might not be able to make progress or might deadlock or whatever. If concurrency is intended as an optimization hook, there's not much risk, but if we are depending on the different threads to make progress in order for the system to work, we need to document that so the app writer knows what the constraints are.

**Comment by tac0turtle** [comment]
This discussion came to mind recently while speaking to a few folks. I am definitely on the side of allowing the app to define their own concurrency model. Would love to see this discussion pick up. Seems @melekes And @alexanderbez have general agreement, but would be cool to see what the new team behind tendermint thinks.
